biojava-core/src/main/java/org/biojava/nbio/core/search/io/blast/BlastXMLQuery.java

```diff
@@ -0,0 +1,270 @@
+ */
+public class BlastXMLQuery implements ResultFactory {
+    private static final Logger log = Logger.getLogger(BlastXMLQuery.class.getName());
+    Document blastDoc = null;
+    private File targetFile;
+    private List<Sequence> queryReferences, databaseReferences;
+    private Map<String,Sequence> queryReferencesMap, databaseReferencesMap;
+    
+    @Deprecated
+    public BlastXMLQuery(String blastFile) throws Exception {
+        readFile(blastFile);
+    }
+    
+    public BlastXMLQuery() {
+        
+    }
+    public void setFile(File f){
+        targetFile = f;
+    }
+    /**
+     * 
+     * @param maxEScore
+     * @return
+     * @throws Exception 
+     */
+    @Deprecated
+    public LinkedHashMap<String, ArrayList<String>> getHitsQueryDef(double maxEScore) throws Exception {
+        LinkedHashMap<String, ArrayList<String>> hitsHashMap = new LinkedHashMap<String, ArrayList<String>>();
+        log.info("Query for hits");
+        ArrayList<Element> elementList = XMLHelper.selectElements(blastDoc.getDocumentElement(), "BlastOutput_iterations/Iteration[Iteration_hits]");
+        log.info(elementList.size() + " hits");
+
+        for (Element element : elementList) {
+            Element iterationquerydefElement = XMLHelper.selectSingleElement(element, "Iteration_query-def");
+            String querydef = iterationquerydefElement.getTextContent();
+            Element iterationHitsElement = XMLHelper.selectSingleElement(element, "Iteration_hits");
+            ArrayList<Element> hitList = XMLHelper.selectElements(iterationHitsElement, "Hit");
+            for (Element hitElement : hitList) {
+                Element hitaccessionElement = XMLHelper.selectSingleElement(hitElement, "Hit_accession");
+                String hitaccession = hitaccessionElement.getTextContent();
+                Element hithspsElement = XMLHelper.selectSingleElement(hitElement, "Hit_hsps");
+                ArrayList<Element> hspList = XMLHelper.selectElements(hithspsElement, "Hsp");
+                for (Element hspElement : hspList) {
+                    Element evalueElement = XMLHelper.selectSingleElement(hspElement, "Hsp_evalue");
+                    String value = evalueElement.getTextContent();
+                    double evalue = Double.parseDouble(value);
+                    if (evalue <= maxEScore) {
+                        ArrayList<String> hits = hitsHashMap.get(querydef);
+                        if (hits == null) {
+                            hits = new ArrayList<String>();
+                            hitsHashMap.put(querydef, hits);
+                        }
+                        hits.add(hitaccession);
+                    }
+                }
+            }
+        }
+
+        return hitsHashMap;
+    }
+    
+    private void readFile(String blastFile) throws Exception{
+        log.info("Start reading " + blastFile);
+        blastDoc = XMLHelper.loadXML(blastFile);
+        log.info("Read finished");
+    }
+    

```

Comment:
Hi Jose,
I share what you say and in fact this was an attempt to do what you assert
in your #111, when you said "throw early, catch late" since there are 3
underlying exceptions there: SAXException, IOException,
ParserConfigurationException and my will was to pass them to the top module
level: SearchIO class.
Anyway this brought me to think and refine this handling because
SAXException and ParserConfigurationException are related to the underlying
xml and can be conceptually grouped in a more generic ParseException to be
forwarded to the top level.
See modifications.

2015-09-22 20:49 GMT+02:00 Jose Manuel Duarte <notifications@github.com>:

> In
> biojava-core/src/main/java/org/biojava/nbio/core/search/io/blast/BlastXMLQuery.java
> <https://github.com/biojava/biojava/pull/329#discussion_r40127534>:
>
> > +                        }
> > +                        hits.add(hitaccession);
> > +                    }
> > +                }
> > +            }
> > +        }
> > +
> > +        return hitsHashMap;
> > +    }
> > +
> > +    private void readFile(String blastFile) throws Exception{
> > +        log.info("Start reading " + blastFile);
> > +        blastDoc = XMLHelper.loadXML(blastFile);
> > +        log.info("Read finished");
> > +    }
> > +
>
> Try to avoid throwing generic Exceptions and be as specific as possible.
> When you catch this one below, you are catching not only the ones thrown by
> loadXML, but also any other potential Exception like a null pointer. That's
> not so bad but it can make debugging quite complicated. We are not very
> consistent with this in biojava yet, but we are trying to improve the
> situation, see #111 <https://github.com/biojava/biojava/issues/111>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biojava/biojava/pull/329/files#r40127534>.
>